core/impl/src/main/java/com/blazebit/persistence/impl/function/jsonset/AbstractJsonFunction.java

```diff
@@ -0,0 +1,124 @@
+        return argument.startsWith("$");
+    }
+
+    /**
+     * Introduce JSON path template as optional first parameter
+     * If first parameter is a literal string, interpret it as JSON path. The remaining arguments are interpreted
+     * as varargs named parameters that are used as positional parameters for the json path template.
+     * For postgres we need to parse the json path template and reformulate it to the postgres syntax.
+     */
+    protected static String toJsonPathTemplate(List<Object> pathElements, int to, boolean quotePathElements) {
+        Object firstArgument = pathElements.get(0);
+        if (firstArgument instanceof String && isJsonPathTemplate((String) firstArgument)) {
+            return (String) firstArgument;
+        } else {
+            StringBuilder jsonPathBuilder = new StringBuilder("$");
+            for (int i = 0; i < to; i++) {
+                Object currentPathElement = pathElements.get(i);
+                if (currentPathElement instanceof Integer) {
+                    jsonPathBuilder.append('[');
+                    jsonPathBuilder.append((int) currentPathElement);
+                    jsonPathBuilder.append(']');
+                } else {
+                    jsonPathBuilder.append('.');
+                    if (quotePathElements) {
+                        jsonPathBuilder.append("\"");
+                    }
+                    jsonPathBuilder.append((String) currentPathElement);
+                    if (quotePathElements) {
+                        jsonPathBuilder.append("\"");
+                    }
+                }
+            }
+            return jsonPathBuilder.toString();
+        }
+    }
+
+    protected static void renderJsonPathTemplate(FunctionRenderContext context, String jsonPathTemplate,
+                                                   int templateParameterOffset) {
+        List<String> concatenationParts = splitByParameterPlaceholder(jsonPathTemplate);
+        if (concatenationParts.size() == 1) {
+            context.addChunk(quoteSingle(jsonPathTemplate));
+        } else {
+            context.addChunk("concat(");
+            context.addChunk(quoteSingle(concatenationParts.get(0)));
+
+            for (int i = 1; i < concatenationParts.size(); i++) {
+                if (i <= concatenationParts.size() - 2) {
+                    context.addChunk(",concat(");
+                    context.addArgument(templateParameterOffset++);
+                    context.addChunk(",concat(");
+                } else if (i <= concatenationParts.size() - 1) {
+                    context.addChunk(",concat(");
+                    context.addArgument(templateParameterOffset++);
+                    context.addChunk(",");
+                }
+                context.addChunk(quoteSingle(concatenationParts.get(i)));
+            }
+            for (int i = 0; i < (concatenationParts.size() - 2) * 2 + 2; i++) {
+                context.addChunk(")");
+            }
```

Comment:
`concat` is not a standard function unfortunately. I would suggest you inject the `ConcatFunction` object for the respective database into the constructor like we do in the various `GroupConcateBased...` classes.

Reply:
Also, wouldn't this be much easier?

```suggestion
            for (int i = 0; i < concatenationParts.size() - 1; i++) {
                context.addChunk("concat(");
                context.addChunk(quoteSingle(concatenationParts.get(i)));
                context.addChunk(",concat(");
                context.addArgument(templateParameterOffset++);
                context.addChunk(",");
            }
            context.addChunk(quoteSingle(concatenationParts.get(i)));
            for (int i = 0; i < (concatenationParts.size() - 2) * 2 + 2; i++) {
                context.addChunk(")");
            }
```

Reply:
Improved